contrib/platform/src/com/sun/jna/platform/linux/Udev.java

```diff
@@ -0,0 +1,547 @@
+         */
+        public String getSysname() {
+            return INSTANCE.udev_device_get_syspath(this);
+        }
+
+        /**
+         * Gets the devnode of this device
+         *
+         * @return a string that describes the devnode. On failure, may return NULL.
+         */
+        public String getDevnode() {
+            return INSTANCE.udev_device_get_devnode(this);
+        }
+
+        /**
+         * Gets the devtype of this device
+         *
+         * @return a string that describes the devtype. On failure, may return NULL.
+         */
+        public String getDevtype() {
+            return INSTANCE.udev_device_get_devtype(this);
+        }
+
+        /**
+         * Gets the subsystem of this device
+         *
+         * @return a string that describes the subsystem. On failure, may return NULL.
+         */
+        public String getSubsystem() {
+            return INSTANCE.udev_device_get_subsystem(this);
+        }
+
+        /**
+         * Retrieves a device attribute from this device
+         *
+         * @param sysattr
+         *            The attribute to retrieve.
+         * @return a string of the requested value. On error, NULL is returned.
+         *         Attributes that may contain NUL bytes should not be retrieved with
+         *         udev_device_get_sysattr_value(); instead, read them directly from the
+         *         files within the device's syspath.
+         */
+        public String getSysattrValue(String sysattr) {
+            return INSTANCE.udev_device_get_sysattr_value(this, sysattr);
+        }
+
+        /**
+         * Retrieves a device property from this device
+         *
+         * @param key
+         *            The property to retrieve.
+         * @return a string of the requested value. On error, NULL is returned.
+         */
+        public String getPropertyValue(String key) {
+            return INSTANCE.udev_device_get_property_value(this, key);
+        }
+    }
+
+    /**
+     * Allocates a new udev context object and returns a pointer to it. This object
+     * is opaque and must not be accessed by the caller via different means than
+     * functions provided by libudev. Initially, the reference count of the context
+     * is 1.
+     *
+     * @return On success, returns a pointer to the allocated udev context. On
+     *         failure, NULL is returned.
+     */
+    UdevContext udev_new();
+
+    /**
+     * Acquire further references to a udev context object.
+     *
+     * @param udev
+     *            A udev context object.
+     * @return the argument that was passed, unmodified.
+     */
+    UdevContext udev_ref(UdevContext udev);
+
+    /**
+     * Drop a reference to a udev context object. Once the reference count hits 0,
+     * the context object is destroyed and freed.
+     *
+     * @param udev
+     *            A udev context object.
+     */
+    void udev_unref(UdevContext udev);
```

Comment:
This is not correct - the definition needs to match the C version. I dropped the return in my suggestion for the OO binding - but only on the java level, not on the Java<->C interface level. Depending on how the return value is returned, memory might need to be reserved for this. If it does not happen, strange issues could result, so please return the UdevContext here.

Reply:
Gotcha, I missed that subtle distinction.